keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,552 @@
+Other controller use informers and caches to reduce the load on API server. The conjob controller does a periodic poll and sweep of all the objects and acts on them. Also CronJob controller has only one worker doing this. 
+
+1. syncs all CronJob objects [every 10 seconds](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cronjob/cronjob_controller.go#L98). 
+2. Using pager library, gets all Pods and all CronJobs and [processes them one by one](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cronjob/cronjob_controller.go#L136)
+
+This is not a scalable design and ends up loading the API server.
+
+#### Proposed rearchitect
+
+With proposed rearchitecture we aim to:
+1. Reduce the potential scale issues when using lots of CronJob objects
+2. Reduce load on API server in such cases.
+3. Reduce memory usage
+
+##### Informers and Caches
+To reduce the need to list all Jobs and CronJobs frequently to reconcile, we propose to replace it with an Informers and WorkQueue based architecture. We would be sharing the [same informer cache](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L450) as the Job controller uses. 
+
+
+```golang
+
+// 
+type CronJobController struct {
+	kubeClient 			clientset.Interface
+	queue	   			workqueue.RateLimitingInterface
+	recorder   			record.EventRecorder
+	
+	jobControl 			jobControlInterface
+	cronJobControl  		cronJobControlInterface
+	
+	jobLister  			batchv1listers.JobLister
+	cronJobLister 			CronJobLister
+	
+	jobListerSynced 		cache.InformerSynced
+	cronJobListerSynced		cache.InformerSynced
+}
+
+// CronJobControllerConfiguration contains controller config
+type CronJobControllerConfiguration struct {
+	// concurrentCronJobSyncs is the number of cronjob objects that are
+	// allowed to sync concurrently. Larger number = more responsive
+	// but more CPU (and network) load.
+	ConcurrentCronJobSyncs int32
+}
+
+
+// Controller code
+
+func (cjc *CronJobController)  Run() {
+	...
+	if !cache.WaitForNamedCacheSync("cronjob", stopCh, cjc.jobListerSynced, cjc.cronJobListerSynced) {
+		return
+	}
+    for i := 0; i < workers; i++ {
+        go wait.Until(cjc.worker, time.Second, stopCh)
+    }
+...
+}
+
+func (cjc *CronJobController) func worker() {
+    for cjc.processNextWorkItem() {}
+}
+
+func (cjc *CronJobController) func processNextWorkItem() {
+	key, quit := cjc.queue.Get()
+	if quit {
+		return false
+	}
+	defer cjc.queue.Done(key)
+	if err := cjc.sync(key.(string)); err != nil {
+		utilruntime.HandleError(fmt.Errorf("Error syncing CronJobController %v, requeuing: %v", key.(string), err))
+		cjc.queue.AddRateLimited(key)
+	} else {
+		cjc.queue.Forget(key)
+	}
+	return true
+}
+
+func (cjc *CronJobController)  sync(cronJobKey) {
+	childrenJobs := cjc.jobLister.GetJobsForCronJob(cronJobKey)
+	// reuse/refactor the existing syncOne function
+}
```

Comment:
Less code and more information, we'll get to the implementation part. Also, the above can be linked from https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md#rough-structure

Reply:
Thanks for the link. Will remove appropriate code sections.